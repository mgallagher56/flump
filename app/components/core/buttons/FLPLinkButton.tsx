import React from 'react';

import type { ButtonProps } from '@chakra-ui/react';
import { Button } from '@chakra-ui/react';
import { Link } from '@remix-run/react';
import * as pressable from '@zag-js/pressable';
import type { PressEvent } from '@zag-js/pressable/dist/pressable.types';
import { normalizeProps, useMachine } from '@zag-js/react';

interface FLPLinkButtonProps extends ButtonProps {
  preventFocusOnPress?: boolean;
  to: string;
  onPress?: (e: PressEvent) => void;
}

const FLPLinkButton: React.FC<React.PropsWithChildren<FLPLinkButtonProps>> = ({
  children,
  colorScheme = 'blue',
  isDisabled,
  preventFocusOnPress,
  to
}) => {
  const [state, send] = useMachine(
    pressable.machine({
      id: 'pressableBaseButton',
      disabled: isDisabled,
      preventFocusOnPress: preventFocusOnPress ?? true
    })
  );

  const api = pressable.connect(state, send, normalizeProps);

  return (
    <Button colorScheme={colorScheme} as={Link} to={to} variant="link" {...api.pressableProps}>
      {children}
    </Button>
  );
};

export default FLPLinkButton;
