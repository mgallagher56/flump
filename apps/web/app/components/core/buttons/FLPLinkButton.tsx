import { type FC, type PropsWithChildren } from 'react';

import type { ButtonProps } from '@chakra-ui/react';
import { Button } from '@chakra-ui/react';
import { Link } from 'react-router';

interface FLPLinkButtonProps extends ButtonProps {
  to: string;
}

const FLPLinkButton: FC<PropsWithChildren<FLPLinkButtonProps>> = ({
  children,
  colorScheme = 'blue',
  disabled,
  variant = 'link',
  to,
  ...rest
}) => {
  return (
    <Button as={Link} colorScheme={colorScheme} disabled={disabled} to={to} variant={variant} {...rest}>
      {children}
    </Button>
  );
};

export default FLPLinkButton;
