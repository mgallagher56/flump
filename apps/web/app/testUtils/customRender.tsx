// create a custom render function with ChakraProvider
import { type ReactNode } from 'react';

import { type RenderOptions, type RenderResult, render as rtlRender } from '@testing-library/react';
import userEvent, { type UserEvent } from '@testing-library/user-event';
import { Provider } from '~/components/ui/provider';

const user = userEvent.setup();

const customRender = (ui: ReactNode, options?: RenderOptions): RenderResult & { user: UserEvent } => {
  const wrappedUi = <Provider>{ui}</Provider>;

  return {
    ...rtlRender(wrappedUi, options),
    user
  };
};

export default customRender;
