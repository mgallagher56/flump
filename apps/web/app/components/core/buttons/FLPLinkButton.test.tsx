import customRender from '~/testUtils/customRender';

import FLPLinkButton from './FLPLinkButton';

describe('FLPLinkButton', () => {
  test('renders as expected', () => {
    const { container, getByText } =customRender(<FLPLinkButton text={'Button'} to="/test" />);
    expect(container).toMatchSnapshot();
    expect(getByText('Button')).toBeDefined();
  });
  test('renders as expected with a custom class', () => {
    const { container, getByText } =customRender(<FLPLinkButton className="test-class" text={'Button'} to="/test" />);
    expect(container).toMatchSnapshot();
    expect(getByText('Button')).toBeDefined();
  });

  test('renders as expected with a colorPalette', () => {
    const { container, getByText } =customRender(<FLPLinkButton colorPalette="red" text={'Button'} to="/test" />);
    expect(container).toMatchSnapshot();
    expect(getByText('Button')).toBeDefined();
  });
});
