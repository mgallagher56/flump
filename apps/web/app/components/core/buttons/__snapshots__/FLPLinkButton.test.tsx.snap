// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FLPLinkButton > renders as expected 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,r,s,u,d,m,l,h)=&gt;
    {let c=document.documentElement,v=["light","dark"];function p(i){(Array.isArray(e)?e:[e]).forEach(y=&gt;
    {let k=y==="class",S=k&&m?d.map(f=&gt;
    m[f]||f):d;k?(c.classList.remove(...S),c.classList.add(i)):c.setAttribute(y,i)}),R(i)}function R(i){h&&v.includes(i)&&(c.style.colorScheme=i)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let i=localStorage.getItem(r)||s,y=l&&i==="system"?a():i;p(y)}catch(i){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  Button
</div>
`;

exports[`FLPLinkButton > renders as expected with a colorPalette 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,r,s,u,d,m,l,h)=&gt;
    {let c=document.documentElement,v=["light","dark"];function p(i){(Array.isArray(e)?e:[e]).forEach(y=&gt;
    {let k=y==="class",S=k&&m?d.map(f=&gt;
    m[f]||f):d;k?(c.classList.remove(...S),c.classList.add(i)):c.setAttribute(y,i)}),R(i)}function R(i){h&&v.includes(i)&&(c.style.colorScheme=i)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let i=localStorage.getItem(r)||s,y=l&&i==="system"?a():i;p(y)}catch(i){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  Button
</div>
`;

exports[`FLPLinkButton > renders as expected with a custom class 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,r,s,u,d,m,l,h)=&gt;
    {let c=document.documentElement,v=["light","dark"];function p(i){(Array.isArray(e)?e:[e]).forEach(y=&gt;
    {let k=y==="class",S=k&&m?d.map(f=&gt;
    m[f]||f):d;k?(c.classList.remove(...S),c.classList.add(i)):c.setAttribute(y,i)}),R(i)}function R(i){h&&v.includes(i)&&(c.style.colorScheme=i)}function a(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let i=localStorage.getItem(r)||s,y=l&&i==="system"?a():i;p(y)}catch(i){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  Button
</div>
`;
